src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,595 @@
+  else (
+    add_char id buf;
+    write_U32_of_int (Buffer.length section_buf) buf;
+    Buffer.add_buffer buf section_buf )
+
+(* type: section 1 *)
+let type_encode func buf =
+  vector_encode_from_namedvalues ~rev_values:true func
+    (fun v ->
+      let v = Indexed.get v in
+      add_char '\x60' buf;
+      match v with
+      | Runtime.Local { type_f = t; locals = _; body = _; id = _ } -> begin
+        match t with
+        | Bt_raw (_ind, (pt, rt)) ->
+          fun buf ->
+            paramtype_encode pt buf;
+            resulttype_encode rt buf
+        | _ -> assert false
+      end
+      | _ -> fun _ -> () )
+    buf
+
+(* function: section 3 *)
+let function_encode func buf =
+  vector_encode_from_namedvalues ~rev_values:true func
+    (fun v ->
+      let i = Indexed.get_index v in
+      write_U32_of_int i )
+    buf
+
+(* table: section 4 *)
+let tables_encode table buf =
+  vector_encode_from_namedvalues ~rev_values:true table
+    (fun v ->
+      let v = Indexed.get v in
+      match v with
+      | Runtime.Local table -> table_encode table
+      | _ -> fun _ -> () )
+    buf
+
+(* memory: section 5 *)
+let memories_encode memories buf =
+  vector_encode_from_namedvalues ~rev_values:true memories
+    (fun v ->
+      let v = Indexed.get v in
+      match v with
+      | Runtime.Local memory -> memory_encode memory
+      | _ -> fun _ -> () )
+    buf
+
+(* global: section 6 *)
+let globals_encode globals buf =
+  vector_encode_from_namedvalues ~rev_values:false globals
+    (fun v ->
+      let v = Indexed.get v in
+      match v with
+      | Runtime.Local global -> global_encode global buf
+      | _ -> fun _ -> () )
+    buf
+
+(* start: section 8 *)
+let start_encode int_opt buf =
+  match int_opt with None -> () | Some funcidx -> write_U32_of_int funcidx buf
+
+let code_encode func buf =
+  vector_encode_from_namedvalues ~rev_values:true func
+    (fun v ->
+      let v = Indexed.get v in
+      let code_buf = Buffer.create 16 in
+      match v with
+      | Runtime.Local { type_f; locals; body; id = _ } ->
+        locals_encode type_f locals code_buf;
+        expr_encode body code_buf;
+        write_U32_of_int (Buffer.length code_buf) buf;
+        fun buf -> Buffer.add_buffer buf code_buf
+      | _ -> fun _ -> () )
+    buf
+
+let encode (modul : Binary.modul) =
+  let buf = Buffer.create 256 in
+  Buffer.add_string buf magic;
+  Buffer.add_string buf version;
+  section_encode '\x01' type_encode modul.func buf;
+  section_encode '\x03' function_encode modul.func buf;
+  section_encode '\x04' tables_encode modul.table buf;
+  section_encode '\x05' memories_encode modul.mem buf;
+  section_encode '\x06' globals_encode modul.global buf;
+  section_encode '\x08' start_encode modul.start buf;
+  section_encode '\x0A' code_encode modul.func buf;
+
+  (* export_encode modul.exports buf; *)
+  (* element_encode modul.elem buf; *)
+  (* data_encode modul.data buf; *)
+  (* data count *)
```

Comment:
Be careful, the order of sections order is not exactly the order of the identifier.

They go: 0x01, 0x02, 0x03, ..., 0x09, 0x0C (datacount), 0x0A (code), 0x0B (data).